Firestore/core/src/firebase/firestore/remote/grpc_stream.cc

```diff
@@ -173,22 +179,28 @@ void GrpcStream::FinishAndNotify(const Status& status) {
 }
 
 void GrpcStream::Shutdown() {
+  LOG_DEBUG(
+      "GrpcStream('%s'): shutting down; completion(s): %s, is finished: %s",
+      this, completions_.size(), is_grpc_call_finished_);
+
   MaybeUnregister();
-  if (completions_.empty()) {
-    // Nothing to cancel -- either the call was already finished, or it has
-    // never been started.
-    return;
+
+  if (!completions_.empty() && !is_grpc_call_finished_) {
```

Comment:
I don't entirely understand why both conditions must hold here. It seems as if `is_grpc_call_finished_` should be sufficient, no?

In particular this suggests that it's possible for e.g. `completions_` to be empty but `is_grpc_call_finished_` to be false and we wouldn't want to call `FinishGrpcCall`, but that seems wrong.

Reply:
> this suggests that it's possible for e.g. completions_ to be empty but is_grpc_call_finished_ to be false and we wouldn't want to call FinishGrpcCall, but that seems wrong.

The stream is in this state before it is started. Finishing a non-started stream would also fail a gRPC assertion. I don't think we should ever run into the case when `Stream` tries to `Finish` a `GrpcStream` before it is started. I'm mainly doing this because it's an easy check to make.

If you think protecting against this case isn't worthwhile, I'll simplify the check. Otherwise, I'll add a comment.

Reply:
OK. Looking at the prior state more closely there was a comment there that explained this and it made sense. It's possible you may want to merge something of that comment with what you have below.